### DISCUSSION TITLE
Creating a tar.gz with respect to symlinks

### DISCUSSION THREAD

**Discussion by TylerMizuyabu** [icsharpcode/SharpZipLib#830] [Q&A]
Hello, I'm currently trying to use this library to create a .tar.gz of a user's local workspace. This workspace may or may not contain symlinks, and ideally I want to preserve all symlinks that link to folder/files within the local workspace. Currently when I try to do this with code like the following:
```
FileStream outStream = File.Create("workspace.tar.gz");
Stream gzoStream = new GZipOutputStream(outStream);
TarArchive tarArchive = TarArchive.CreateOutputTarArchive(gzoStream);
foreach (FileInfo info in fileInfos)
{
		tarArchive.WriteEntry(TarEntry.CreateEntryFromFile(info.FullName), false);
}
```

The symlink file does not get preserved. Instead one of the following two things happen:
1. If the symlink targets a file, I get an error because the size field within the header does not match the actual size of the targeted file.
2. If the symlink targets a directory, the target directory gets added to the tarArchive but not the symlink. 

Is there a way I can add the symlink itself to the archive?

**Comment by piksel** [comment]
Yes, but you would need to identify the link yourself (~no built-in way to detect symlinks in .NET if I recall~ see edit) and add an entry using something like this:

```cs
foreach (FileInfo info in fileInfos)
{
	if (isSymlink(info, out var target)) 
    {
        var tarEntry = TarEntry.CreateTarEntry(info.FullName);
        tarEntry.TarHeader.LinkName = target;
        tarEntry.TarHeader.TypeFlag = TarHeader.LF_SYMLINK;
        tarArchive.WriteEntry(tarEntry, false);
    } 
    else 
    {
		tarArchive.WriteEntry(TarEntry.CreateEntryFromFile(info.FullName), false);
    }
}
```

**Edit**: Actually, since .NET 6 it is now possible to work with symlinks and unix permissions, so perhaps support for this ought to be (finally) added to `TarArchive`.

**Reply by TylerMizuyabu** [reply]
Added an additional call to `entry.GetFileTarHeader(entry.TarHeader, info.FullName);` because the WriteEntry function was complaining that the tar entry's `File` property was null and it works well for symlinks that target directories. However it still produces the error `request to write '69' bytes exceeds size in header of '10' bytes (Parameter 'count')` for symlinks that target files.

I'm guessing the the `file.OpenRead` call in the tar archive's `WriteEntryCore` method is opening up a stream to the target file so there is still a mismatch between the length it expects and the length of the content it is reading from. Is there additional configuration I am missing or perhaps the additional call to `GetFileTarHeader` was incorrect?

**Reply by piksel** [reply]
Perhaps you can clear the file size manually using `tarEntry.TarHeader.Size` (or equivalent property, I'm on mobile right now). Otherwise this might actually not be possible. I will take a look as soon as I get time!
